File: fixtures/programming/haskell/TaskStore.hs
-- | Task management module — pure functional approach.
module TaskStore
  ( Task(..)
  , TaskStore
  , empty
  , addTask
  , completeTask
  , pendingTasks
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

-- | A task with an ID, title, and completion flag.
data Task = Task
  { taskId    :: Int
  , taskTitle :: String
  , taskDone  :: Bool
  } deriving (Show, Eq)

-- | In-memory store for tasks.
data TaskStore = TaskStore
  { storeTasks :: Map Int Task
  , storeNextId :: Int
  } deriving (Show)

-- | Creates an empty store.
empty :: TaskStore
empty = TaskStore Map.empty 1

-- | Adds a task and returns the updated store with the new task.
addTask :: String -> TaskStore -> (Task, TaskStore)
addTask title store =
  let task = Task (storeNextId store) title False
      newStore = store
        { storeTasks = Map.insert (taskId task) task (storeTasks store)
        , storeNextId = storeNextId store + 1
        }
  in (task, newStore)

-- | Marks a task as done.
completeTask :: Int -> TaskStore -> TaskStore
completeTask tid store =
  store { storeTasks = Map.adjust (\t -> t { taskDone = True }) tid (storeTasks store) }

-- | Returns all pending tasks.
pendingTasks :: TaskStore -> [Task]
pendingTasks = filter (not . taskDone) . Map.elems . storeTasks

File: fixtures/programming/ocaml/task_store.ml
(** Task management module — functional approach with immutable data. *)

(** Status of a task. *)
type status = Pending | Done

(** A task with ID, title, and status. *)
type task = {
  id : int;
  title : string;
  status : status;
}

(** In-memory task store. *)
type store = {
  tasks : task list;
  next_id : int;
}

(** Creates an empty store. *)
let empty : store = { tasks = []; next_id = 1 }

(** Adds a task with the given title. *)
let add_task title store =
  let task = { id = store.next_id; title; status = Pending } in
  let new_store = { tasks = task :: store.tasks; next_id = store.next_id + 1 } in
  (task, new_store)

(** Marks a task as done by ID. *)
let complete_task id store =
  let update t = if t.id = id then { t with status = Done } else t in
  { store with tasks = List.map update store.tasks }

(** Returns all pending tasks. *)
let pending_tasks store =
  List.filter (fun t -> t.status = Pending) store.tasks
